core/src/main/codegen/templates/Parser.jj

```diff
@@ -6271,70 +6271,61 @@ SqlNode BuiltinFunctionCall() :
             s = span();
         }
         <LPAREN>
+        /**
+        * We want to accept:
+        * [BOTH/TRAILING/LEADING] X from Y (trimChars = X, exp = Y)
+        * X (exp = X)
+        * X, Y (Equivalent to trimChars=Y, exp=X)
+        * NOTE: SqlStdOperatorTable.TRIM expects values in the order trimchars, exp
+        */
+        [
+            <BOTH> {
+                s.add(this);
+                flag = SqlTrimFunction.Flag.BOTH.symbol(getPos());
+            }
+        |
+            <TRAILING> {
+                s.add(this);
+                flag = SqlTrimFunction.Flag.TRAILING.symbol(getPos());
+            }
+        |
+            <LEADING> {
+                s.add(this);
+                flag = SqlTrimFunction.Flag.LEADING.symbol(getPos());
+            }
+        ]
+        // Distinguishing the "[x] FROM y" syntax at this point would require expensive and potentially
+        // tricky syntactic lookahead. So instead, we opt to manually throw contextExceptions
+        [ e = Expression(ExprContext.ACCEPT_SUB_QUERY) ]
         [
-            LOOKAHEAD(2)
-            [
-                <BOTH> {
-                    s.add(this);
-                    flag = SqlTrimFunction.Flag.BOTH.symbol(getPos());
-                }
-            |
-                <TRAILING> {
-                    s.add(this);
-                    flag = SqlTrimFunction.Flag.TRAILING.symbol(getPos());
-                }
-            |
-                <LEADING> {
-                    s.add(this);
-                    flag = SqlTrimFunction.Flag.LEADING.symbol(getPos());
-                }
-            ]
-            [ trimChars = Expression(ExprContext.ACCEPT_SUB_QUERY) ]
```

Comment:
The original issue was that from non single token trim queries, we always take this path. I can explain if needed.